src/transport/raw/TCP.cpp

```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+        // established
+        tcp->HandleConnectionComplete(activeConnection, CHIP_NO_ERROR);
     }
     else
     {
-        bool connectionStored = false;
-        for (size_t i = 0; i < tcp->mActiveConnectionsSize; i++)
-        {
-            if (!tcp->mActiveConnections[i].InUse())
-            {
-                tcp->mActiveConnections[i].Init(endPoint);
-                connectionStored = true;
-                break;
-            }
-        }
-
-        // since we track end points counts, we always expect to store the
-        // connection.
-        if (!connectionStored)
-        {
-            endPoint->Free();
-            ChipLogError(Inet, "Internal logic error: insufficient space to store active connection");
-        }
+        ChipLogError(Inet, "Connection establishment with [%s:%d] encountered an error: %s", ipAddrStr, port, ErrorStr(err));
+        endPoint->Free();
+        tcp->mUsedEndPointCount--;
     }
 }
 
-void TCPBase::OnConnectionClosed(Inet::TCPEndPoint * endPoint, CHIP_ERROR err)
+void TCPBase::HandleTCPEndPointConnectionClosed(Inet::TCPEndPoint * endPoint, CHIP_ERROR err)
 {
-    TCPBase * tcp = reinterpret_cast<TCPBase *>(endPoint->mAppState);
+    TCPBase * tcp                            = reinterpret_cast<TCPBase *>(endPoint->mAppState);
+    ActiveConnectionState * activeConnection = tcp->FindActiveConnection(endPoint);
+    Inet::IPAddress ipAddress;
+    uint16_t port;
+    char ipAddrStr[64];
+
+    endPoint->GetPeerInfo(&ipAddress, &port);
+    ipAddress.ToString(ipAddrStr, sizeof(ipAddrStr));
+
+    if (activeConnection == nullptr)
+    {
+        endPoint->Free();
+        return;
+    }
+
+    if (err == CHIP_NO_ERROR && activeConnection->mConnectionState == State::kConnected)
+    {
+        err = CHIP_ERROR_CONNECTION_CLOSED_UNEXPECTEDLY;
+    }
 
-    ChipLogProgress(Inet, "Connection closed.");
+    tcp->CloseConnectionInternal(activeConnection, err, SuppressCallback::No);
 
-    ChipLogProgress(Inet, "Freeing closed connection.");
-    tcp->ReleaseActiveConnection(endPoint);
+    ChipLogProgress(Inet, "Connection closed with [%s:%d]", ipAddrStr, port);
 }
 
-void TCPBase::OnConnectionReceived(Inet::TCPEndPoint * listenEndPoint, Inet::TCPEndPoint * endPoint,
-                                   const Inet::IPAddress & peerAddress, uint16_t peerPort)
+// Handler for incoming connection requests from peer nodes
+void TCPBase::HandleIncomingConnection(Inet::TCPEndPoint * listenEndPoint, Inet::TCPEndPoint * endPoint,
+                                       const Inet::IPAddress & peerAddress, uint16_t peerPort)
 {
-    TCPBase * tcp = reinterpret_cast<TCPBase *>(listenEndPoint->mAppState);
+    TCPBase * tcp                            = reinterpret_cast<TCPBase *>(listenEndPoint->mAppState);
+    ActiveConnectionState * activeConnection = nullptr;
+    Inet::InterfaceId interfaceId;
+    Inet::IPAddress ipAddress;
+    uint16_t port;
+    char ipAddrStr[64];
```

Comment:
See comments above, and for ToString down below.

Also, why is this not declared where it's used?

Reply:
Moved the declaration to where it is used.